src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,175 @@
+        while (element.next != null) {
+            element = element.next;
+        }
+        Node<T> newNode = new Node<T>(element, value, null);
+        element.next = newNode;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
+        if (first == null) {
+            add(value);
+            return;
+        }
+        Node<T> currentNode = getNodeByIndex(index);
 
+        if (currentNode == null) {
+            Node<T> prevCurrent = getNodeByIndex(index - 1);
+            if (prevCurrent == null) {
+                throw new IndexOutOfBoundsException("Index is bigger than list size");
+            }
+            add(value);
+            return;
+        }
+
+        Node<T> prevCurrent = currentNode.prev;
+        Node<T> newNode = new Node<>(prevCurrent, value, currentNode);
+        if (prevCurrent != null) {
+            prevCurrent.next = newNode;
+        } else {
+            first = newNode;
+        }
+        currentNode.prev = newNode;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        if (list == null || list.size() == 0) {
+            throw new IllegalArgumentException("List is empty!");
+        }
+        for (T element : list) {
+            add(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> newNode = getNodeByIndex(index);
+        if (newNode == null) {
+            throw new IndexOutOfBoundsException("Index is bigger, than list size");
+        }
+        return newNode.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> newNode = getNodeByIndex(index);
+        if (newNode == null) {
+            throw new IndexOutOfBoundsException("Index was " + index + " and the size is "
+                    + size());
+        }
+        T oldValue = newNode.item;
+        newNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        var currentNode = getNodeByIndex(index);
+        if (currentNode == null) {
+            throw new IndexOutOfBoundsException("Index was " + index + " and the size is "
+                    + size());
+        }
+        if (index == 0) {
+            var nextCurrent = currentNode.next;
+            if (nextCurrent != null) {
+                nextCurrent.prev = null;
+            }
+            first = nextCurrent;
+            return currentNode.item;
+        }
+        if (currentNode.next == null) {
+            var prevCurrent = currentNode.prev;
+            prevCurrent.next = null;
+            return currentNode.item;
+        }
+        var prevCurrent = currentNode.prev;
+        prevCurrent.next = currentNode.next;
+        var nextCurrent = currentNode.next;
+        nextCurrent.prev = currentNode.prev;
```

Comment:
The logic where you change the relationship between the nodes is better to make a separate method - unlink (Node <T> node)